src/d3a/d3a_core/sim_results/kpi.py

```diff
@@ -0,0 +1,145 @@
+                self.consumer_list.append(child.name)
+                self.consumer_area_list.append(child.parent)
+                self.total_energy_demanded_wh += child.strategy.state.total_energy_demanded_wh
+            elif _is_prosumer_node(child):
+                self.ess_list.append(child.name)
+            elif isinstance(child.strategy, InfiniteBusStrategy):
+                self.buffer_list.append(child.name)
+            if child.children:
+                self._accumulate_devices(child)
+
+    def _accumulate_self_production(self, trade):
+        if trade.seller_origin in self.producer_list:
+            self.total_energy_produced_wh += trade.offer.energy * 1000
+
+    def _accumulate_self_consumption(self, trade):
+        if trade.seller_origin in self.producer_list and trade.buyer_origin in self.consumer_list:
+            self.total_self_consumption_wh += trade.offer.energy * 1000
+
+    def _accumulate_self_consumption_buffer(self, trade):
+        if trade.seller_origin in self.producer_list and trade.buyer_origin in self.ess_list:
+            self.self_consumption_buffer_wh += trade.offer.energy * 1000
+
+    def _dissipate_self_consumption_buffer(self, trade):
+        if trade.seller_origin in self.ess_list:
+            # self_consumption_buffer needs to be exhausted to total_self_consumption
+            # if sold to internal consumer
+            if trade.buyer_origin in self.consumer_list and self.self_consumption_buffer_wh > 0:
+                if (self.self_consumption_buffer_wh - trade.offer.energy * 1000) > 0:
+                    self.self_consumption_buffer_wh -= trade.offer.energy * 1000
+                    self.total_self_consumption_wh += trade.offer.energy * 1000
+                else:
+                    self.total_self_consumption_wh += self.self_consumption_buffer_wh
+                    self.self_consumption_buffer_wh = 0
+            # self_consumption_buffer needs to be exhausted if sold to any external agent
+            elif trade.buyer_origin not in [*self.ess_list, *self.consumer_list] and \
+                    self.self_consumption_buffer_wh > 0:
+                if (self.self_consumption_buffer_wh - trade.offer.energy * 1000) > 0:
+                    self.self_consumption_buffer_wh -= trade.offer.energy * 1000
+                else:
+                    self.self_consumption_buffer_wh = 0
+
+    def _accumulate_infinite_consumption(self, trade):
+        if trade.seller_origin in self.buffer_list and trade.buyer_origin in self.consumer_list:
+            self.total_self_consumption_wh += trade.offer.energy * 1000
+
+    def _dissipate_infinite_consumption(self, trade):
+        if trade.buyer_origin in self.buffer_list and trade.seller_origin in self.producer_list:
+            self.total_self_consumption_wh += trade.offer.energy * 1000
+
+    def _accumulate_energy_trace(self):
+        for c_area in self.consumer_area_list:
+            for market in c_area.past_markets:
```

Comment:
Are you sure the code is using all past markets ? I see that this function is used only when exporting the csv files. During that time, the area.past_markets list will include only the very last market (except if the value of KEEP_PAST_MARKETS is set to True). In case I did not understand something wrong let me know, but I have the impression that this stat should either be incrementally calculated or not exposed to the d3a-web/UI.

Reply:
Yes, you are right, I didn't any effort into dealing with it as part of this story. Since the KPI requirements were constantly changing. 

Since, now they seems to have stabilised and this story has been dragged a lot, I would propose to not do it now and create a small story to deal with it later. 

Let me know if you agree?

Reply:
Sure, just create a new story, as long as it is tracked it is fine by me.